tiledb/sm/query/sparse_global_order_reader.h

```diff
@@ -141,61 +131,122 @@ class SparseGlobalOrderReader : public SparseIndexReaderBase,
   /** Memory budget per fragment for qc tiles. */
   double per_fragment_qc_memory_;
 
-  /** Memory used for result cell slabs. */
-  uint64_t memory_used_rcs_;
-
-  /** How much of the memory budget is reserved for result cell slabs. */
-  double memory_budget_ratio_rcs_;
-
   /* ********************************* */
   /*           PRIVATE METHODS         */
   /* ********************************* */
 
   /** Add a result tile to process, making sure maximum budget is respected. */
-  Status add_result_tile(
+  std::tuple<Status, std::optional<bool>> add_result_tile(
       const unsigned dim_num,
       const uint64_t memory_budget_coords_tiles,
       const uint64_t memory_budget_qc_tiles,
       const unsigned f,
       const uint64_t t,
-      const ArraySchema* const array_schema,
-      bool* budget_exceeded);
-
-  /** corrects memory usage after de-serialization. */
-  Status fix_memory_usage_after_serialization();
+      const ArraySchema* const array_schema);
 
   /** Create the result tiles. */
-  Status create_result_tiles(bool* tiles_found);
+  std::tuple<Status, std::optional<bool>> create_result_tiles();
 
   /** Populate a result cell slab to process. */
-  Status compute_result_cell_slab();
+  Status compute_result_cell_slab(
+      std::vector<ResultCellSlab>& result_cell_slabs);
 
   /**
    * Add a new tile to the queue of tiles currently being processed
-   *  for a specific fragment.
+   * for a specific fragment.
    */
   template <class T>
-  Status add_next_tile_to_queue(
-      bool subarray_set,
+  std::tuple<Status, std::optional<bool>> add_next_tile_to_queue(
       unsigned int frag_idx,
       uint64_t cell_idx,
       std::vector<std::list<ResultTileWithBitmap<uint8_t>>::iterator>&
           result_tiles_it,
-      std::vector<bool>& result_tile_used,
+      std::vector<uint8_t>& result_tile_used,
       std::priority_queue<ResultCoords, std::vector<ResultCoords>, T>&
           tile_queue,
-      std::mutex& tile_queue_mutex,
-      T& cmp,
-      bool* need_more_tiles);
+      std::mutex& tile_queue_mutex);
 
-  /** Computes a tile's Hilbert values, stores them in the comparator. */
-  template <class T>
-  Status calculate_hilbert_values(
-      bool subarray_set, ResultTileWithBitmap<uint8_t>* tile, T& cmp);
+  /** Computes a tile's Hilbert values for a tile. */
+  Status compute_hilbert_values(std::vector<ResultTile*>& result_tiles);
 
   /** Compute the result cell slabs once tiles are loaded. */
   template <class T>
-  Status merge_result_cell_slabs(uint64_t memory_budget, T cmp);
+  Status merge_result_cell_slabs(
+      uint64_t num_cells,
+      std::vector<ResultCellSlab>& result_cell_slabs,
+      T cmp);
+
+  /** Compute parallelization parameters for a tile copy operation. */
```

Comment:
Shall we document all these functions a bit better? We have been diligent with documenting the functionality and all parameters, I suggest we keep it.